services/placement/algo/helper.go

```diff
@@ -416,13 +512,71 @@ func copyPlacement(p services.ServicePlacement) services.ServicePlacement {
 func copyInstances(instances []services.PlacementInstance) []services.PlacementInstance {
 	copied := make([]services.PlacementInstance, len(instances))
 	for i, instance := range instances {
-		copied[i] = placement.NewInstance().
-			SetID(instance.ID()).
-			SetRack(instance.Rack()).
-			SetZone(instance.Zone()).
-			SetWeight(instance.Weight()).
-			SetEndpoint(instance.Endpoint()).
-			SetShards(shard.NewShardsWithIDs(instance.Shards().ShardIDs()))
+		copied[i] = copyInstance(instance)
 	}
 	return copied
 }
+
+func copyInstance(instance services.PlacementInstance) services.PlacementInstance {
+	return placement.NewInstance().
+		SetID(instance.ID()).
+		SetRack(instance.Rack()).
+		SetZone(instance.Zone()).
+		SetWeight(instance.Weight()).
+		SetEndpoint(instance.Endpoint()).
+		SetShards(copyShards(instance.Shards()))
+}
+
+func copyShards(shards shard.Shards) shard.Shards {
```

Comment:
Perhaps `cloneShards` is a better name?  Copy named methods usually take a destination and a source.